### PR TITLE
Best effort co-concretization (iterative algorithm)

### PR DESCRIPTION
**Pull request by alalazo**
**Use case description copied from #27964**

Currently, environments can either be concretized fully together or fully separately. This works well for users who create environments for interoperable software and can use concretization: together. It does not allow environments with conflicting software to be concretized for maximal interoperability.

The primary use-case for this is facilities providing system software. Facilities provide multiple MPI implementations, but all software built against a given MPI ought to be interoperable.

This PR adds a concretization option together_where_possible. When this option is used, Spack will concretize specs in the environment separately, but will optimize for minimal differences in overlapping packages.

**Implementation differences**

The algorithm used here is greedy, since specs are computed in multiple rounds where clingo concretizes together as many input specs as possible. The gist of the algorithm is:

https://github.com/alalazo/spack/blob/f6420eaf6eb5e10a3e4d0b0a1a6c20e4404f92c1/lib/spack/spack/solver/asp.py#L2092-L2135

To relax the requirements on input specs and allow for not solving some of them we add a new indirection:

https://github.com/alalazo/spack/blob/f6420eaf6eb5e10a3e4d0b0a1a6c20e4404f92c1/lib/spack/spack/solver/asp.py#L1787-L1802

and we give clingo the choice to solve for the input literal or not. All the input specs computed in previous rounds are then reused in later rounds to ensure the environment is as contained as possible.

**Comparison with #27964**
When trying out the encoding in #27964 for production environments we saw that there were huge requests:
- On memory during grounding, see https://github.com/spack/spack/pull/27964#issuecomment-1032474892
- On time, during solve, see https://github.com/spack/spack/pull/27964#discussion_r805136646

The solve-time can very likely be improved a lot by removing the symmetry of the new encoding with some artificial rules. The memory requests during the grounding phase instead will necessarily grow linearly with the number of PSIDs needed to solve the environment. That number is not known a-priori. It is capped by the number of specs in an environment (which is computationally unfeasible) but practically is usually much less. For instance, `e4s` has 119 specs and needs 3 process spaces to concretize.

This PR instead takes a greedy approach and perform N solves each with a single process space. It might give a different (and "suboptimal" i.e. with a number of subprocesses greater than strictly needed) result if compared with the encoding in #27964 but it is complete, i.e. if there is at least a solution the algorithm will not give a false negative. On the bright side, due to its greedy nature, it's much faster when solving for environments (`e4s` is solved in 3 rounds in ~1 min.).

### CODE REVIEW

**Comment by alalazo** [comment]
@becker33 @tgamblin [e4s-pr-generate](https://gitlab.spack.io/spack/spack/-/jobs/1739708) took just 2 mins and a single core using `together_where_possible` compared to [3mins. and 16 cores](https://gitlab.spack.io/spack/spack/-/jobs/1738770) when we concretize separately.

File: lib/spack/spack/cmd/solve.py
Comment (becker33) [inline lib/spack/spack/cmd/solve.py]: I think this option should be named after its effects, not its implementation. I don't love `--allow-differences`, but that's the direction I think we should go. The fact it's done in multiple round is an implementation detail.

File: lib/spack/spack/cmd/solve.py
Comment (becker33) [inline lib/spack/spack/cmd/solve.py]: I think we should only show the round information when `spack solve --show output`

File: lib/spack/spack/solver/concretize.lp
Comment (becker33) [inline lib/spack/spack/solver/concretize.lp]:
```suggestion
% each virtual must have at most one provider in the dag
```

File: lib/spack/spack/test/concretize.py
Comment (becker33) [inline lib/spack/spack/test/concretize.py]: I don't understand why this requires three versions of libelf -- `libdwarf@20111030` should be able to reuse either of the other libelf implementations. Is the issue that the first round of concretization is concretizing `libdwarf@20111030 ^ libefl@0.8.13` so we don't back out of that decision?

File: lib/spack/spack/solver/asp.py
Comment (becker33) [inline lib/spack/spack/solver/asp.py]:
```suggestion
        if self._unsolved_specs is None:
```
That way we don't recompute every time if we already know the list is empty.

File: lib/spack/spack/solver/asp.py
Comment (becker33) [inline lib/spack/spack/solver/asp.py]: I don't think the same applies here because the only way to have an empty list is if it's unsat, but we might as well just in case.
```suggestion
        if self._concrete_specs_by_input is None:
```

File: lib/spack/spack/solver/asp.py
Comment (becker33) [inline lib/spack/spack/solver/asp.py]: Same here
```suggestion
        if self._concrete_specs is None:
```

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/spack/cmd/solve.py
Comment (alalazo) [reply lib/spack/spack/cmd/solve.py]: The default when concretizing together is to show also the optimization weights and those change at each round. So here I'm showing the same information + the specs that will be moved to the next round.

File: lib/spack/spack/cmd/solve.py
Comment (alalazo) [reply lib/spack/spack/cmd/solve.py]: Maybe we can just call it `-T, --together-where-possible` to match the concretization mode in environments?

File: lib/spack/spack/test/concretize.py
Comment (alalazo) [reply lib/spack/spack/test/concretize.py]: Yes, I'll investigate why that happens though.

File: lib/spack/spack/solver/asp.py
Comment (alalazo) [reply lib/spack/spack/solver/asp.py]: Fixed in 03e27c2

File: lib/spack/spack/solver/asp.py
Comment (alalazo) [reply lib/spack/spack/solver/asp.py]: Fixed in 03e27c2

File: lib/spack/spack/solver/asp.py
Comment (alalazo) [reply lib/spack/spack/solver/asp.py]: Fixed in 03e27c2

File: lib/spack/spack/solver/concretize.lp
Comment (alalazo) [reply lib/spack/spack/solver/concretize.lp]: Fixed in 03e27c2

File: lib/spack/spack/test/concretize.py
Comment (alalazo) [reply lib/spack/spack/test/concretize.py]: There was a bug on what was reused and another when reconstructing reused specs. They should both be fixed in 53a102c

File: lib/spack/spack/cmd/solve.py
Comment (alalazo) [reply lib/spack/spack/cmd/solve.py]: See 5a54a5e

File: lib/spack/spack/cmd/solve.py
Comment (alalazo) [reply lib/spack/spack/cmd/solve.py]: This is not as good as I would like, but see 032bc0f If it seems acceptable we can defer improving the last bit of `spack solve` to a following, smaller PR.

File: lib/spack/spack/cmd/solve.py
Comment (becker33) [inline lib/spack/spack/cmd/solve.py]:
```suggestion
            if not setup_only:
                _process_result(result, show, required_format, kwargs)
```

File: lib/spack/spack/test/concretize.py
Comment (becker33) [inline lib/spack/spack/test/concretize.py:1704]: Could we improve this by having both root literals (the literals we already use) and dependency literals (explicitly mentioned dependencies of roots) and minimizing the unsolved dependency literals as well? I think that would cause us to concretize the most specific specs first, which would increase our odds of a more minimal build.

**Review by becker33 — CHANGES_REQUESTED** [review]
One request, one idea

File: lib/spack/spack/test/concretize.py
Comment (alalazo) [reply lib/spack/spack/test/concretize.py:1704]: Adding a comment, following a discussion with @becker33 yesterday. We can do something very similar to this by adding weights during the setup phase for the literals that are solved. Then we can minimize the sum of the weights for reuse (currently weights are uniform and equal to 1 for each literal spec).

We can experiment with different ways of computing the weights in a separate PR.

File: lib/spack/spack/cmd/solve.py
Comment (alalazo) [reply lib/spack/spack/cmd/solve.py]: Done in 7cebbff

**Comment by alalazo** [comment]
@becker33 Pending unrelated errors in pipelines I think this is done.

**Review by becker33 — DISMISSED** [review]

**Comment by alalazo** [comment]
@becker33 Is this PR waiting further review before being merged?

**Comment by alalazo** [comment]
@becker33  It's a mystery to me how @eugeneswalker (or more likely Github on his behalf) could dismiss your review with a commit that doesn't exist anymore, but can you please :heavy_check_mark: again? 

I rebased the PR to double check that everything is working correctly with #28673 in `develop`.

**Comment by eugeneswalker** [comment]
> @becker33 It's a mystery to me how @eugeneswalker (or more likely Github on his behalf) could dismiss your review with a commit that doesn't exist anymore, but can you please ✔️ again?

What da!?! I definitely didn't mean to do this, didn't even know about this PR. If anyone knows how this could happen...

**Comment by alalazo** [comment]
@tgamblin Rebased and ready for review.

File: lib/spack/spack/solver/concretize.lp
Comment (becker33) [inline lib/spack/spack/solver/concretize.lp:17]: I think we can constrain this in asp to ensure at least one spec is concretized, that will get us better error messages than doing it in python.

```
:- C = #count{ ID : literal_solved(ID) }, C<1.
```

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/spack/solver/concretize.lp
Comment (alalazo) [reply lib/spack/spack/solver/concretize.lp:17]: I agree with that since we now have #30669 Good catch, since this was not pointed out by a conflict during rebase!

Comment (alalazo) [reply lib/spack/spack/solver/concretize.lp:17]: Should be solved in 7c9ab7156fcf6c0ce905ee3f782760a0da0829e3

**Review by becker33 — COMMENTED** [review]
I think this is ready, except that it will have conflicts with #30038 that will need to be resolved.

File: lib/spack/spack/cmd/solve.py
Comment (becker33) [inline lib/spack/spack/cmd/solve.py]: This argument should be removed, we can just rely on the config option now (or make it a config-pushing option, like `--reuse`.

File: lib/spack/spack/environment/environment.py
Comment (becker33) [inline lib/spack/spack/environment/environment.py]:
```suggestion
        try:
            translation = {'together': True, 'separately': False}
            self.concretization = translation[configuration['concretization']]
        except KeyError:
            self.concretization = spack.config.get('concretizer:unify', 'separately')

Comment (becker33) [reply lib/spack/spack/environment/environment.py]: I think this is a much cleaner way of letting the old way take precedence.

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/spack/cmd/solve.py
Comment (alalazo) [reply lib/spack/spack/cmd/solve.py]: Fixed in 08bc55e

File: lib/spack/spack/environment/environment.py
Comment (alalazo) [reply lib/spack/spack/environment/environment.py]: Fixed in d7f254c

**Review by becker33 — APPROVED** [review]

**Pull request merged**
